arrows/core/generate_texture.h

```diff
@@ -0,0 +1,441 @@
+
+  return i1 + (i2 - i1) * normx;
+}
+
+
+/// This functions dispatches calls to templated versions of bilinear interpolation based on pixels type
+/**
+* \param img [in] image
+* \param x [in] location
+* \param y [in] location
+* \param d [in] depth
+*/
+double bilinear_interp_safe(vital::image const& img, double x, double y, int d=0)
+{
+  if (img.pixel_traits() == vital::image_pixel_traits_of<unsigned char>())
+    return bilinear_interp_safe(vital::image_of<unsigned char>(img), x, y, d);
+  else if (img.pixel_traits() == vital::image_pixel_traits_of<int>())
+    return bilinear_interp_safe(vital::image_of<int>(img), x, y, d);
+  else if (img.pixel_traits() == vital::image_pixel_traits_of<float>())
+    return bilinear_interp_safe(vital::image_of<float>(img), x, y, d);
+  else if (img.pixel_traits() == vital::image_pixel_traits_of<double>())
+    return bilinear_interp_safe(vital::image_of<double>(img), x, y, d);
+}
+
+
+/// Base functor used to fuse multiple images. It adjusts the contribution of each image.
+struct image_fusion_method
+{
+  virtual void operator ()(std::vector<double>& scores) const = 0;
+};
+
+
+/// This functor sets the highest score to 1 and the other to 0
+struct select_max_score : image_fusion_method
+{
+  virtual void operator ()(std::vector<double>& scores) const
+  {
+    auto max_it = std::max_element(scores.begin(), scores.end());
+    if (*max_it > 0)
+    {
+      std::fill(scores.begin(), scores.end(), 0.0);
+      *max_it = 1.0;
+    }
+  }
+};
+
+
+/// This functor normalizes the scores
+struct normalize_scores : image_fusion_method
+{
+  virtual void operator ()(std::vector<double>& scores) const
+  {
+    double sum = std::accumulate(scores.begin(), scores.end(), 0.0);
+    if (sum > 0)
+      std::for_each(scores.begin(), scores.end(), [&sum](double& v) { v /= sum; });
+  }
+};
+
+
+/// This function renders a triangle and fills it with data from different images
+/**
+ * \param v1 [in] 2D triangle point
+ * \param v2 [in] 2D triangle point
+ * \param v3 [in] 2D triangle point
+ * \param pt1 [in] 3D triangle point
+ * \param pt2 [in] 3D triangle point
+ * \param pt3 [in] 3D triangle point
+ * \param cameras [in] list of cameras
+ * \param images [in] list of images
+ * \param depths_pt1 [in] depths of pt1 w.r.t cameras
+ * \param depths_pt2 [in] depths of pt2 w.r.t cameras
+ * \param depths_pt3 [in] depths of pt3 w.r.t cameras
+ * \param depth_maps [in] list of depthmaps
+ * \param texture [out] texture where the triangle is rendered
+ * \param depth_threshold [in] threshold used for depth test
+ */
+template <class T>
+void render_triangle_from_images(vital::vector_2d const& v1, vital::vector_2d const& v2, vital::vector_2d const& v3,
+                                 vital::vector_3d const& pt1, vital::vector_3d const& pt2, vital::vector_3d const& pt3,
+                                 std::vector<vital::camera_sptr> const& cameras, std::vector<vital::image> const& images,
```

Comment:
From my point of view, we can force the input and output images to have the same type.
I allowed different types mainly for convenience because the satellite images that we used in core3D were `uint16` and the output texture needed to be `uchar` to be directly compatible with any viewer.

`image_of` also forces the different input images to have the same pixel type, which is important.